File: goopylib/maths/packing/shelf.py
"""
Module defining rectangle packing using shelf-based algorithms.

See https://pds25.egloos.com/pds/201504/21/98/RectangleBinPack.pdf

"""

from __future__ import annotations

from typing import Optional, Callable

from .packing import PackingAlgorithm, SortingFunction, Item, Bin, ShelvedBin, sort_by_long_side, sort_by_short_side


# Bin shadows the Python built-in 'bin', but until I can think
# of a better name to call it, I'm going to stick with bin.
# pylint: disable=redefined-builtin


class Shelf:
    """
    Class representing a single shelf (row) for shelf-based packing algorithms

    Args:
        vertical_offset: the vertical position of the shelf
        bin: the bin associated with the shelf
    """

    def __init__(self, vertical_offset: float, bin: ShelvedBin) -> None:
        """
        Class representing a single shelf (row) for shelf-based packing algorithms

        Args:
            vertical_offset: the vertical position of the shelf
            bin: the bin associated with the shelf
        """

        self.width: float = bin.width
        self.height: float = 0

        self.packed_width: float = 0
        self.available_width: float = self.width

        self.vertical_offset: float = vertical_offset

        self.is_open = True
        self.bin: Bin = bin
        self.items: list[Item] = []

    def __repr__(self) -> str:
        return f"Shelf(offset={self.vertical_offset}, packed={self.packed_width}) with {len(self.items)} items"

    def _fits(self, item: Item) -> bool:
        """
        Returns:
            if an item fits inside this shelf.
        """
        if self.is_open:
            return item.width <= self.available_width and self.vertical_offset + item.height <= self.bin.height
        else:
            return item.width <= self.available_width and item.height <= self.height

    def _fits_above(self, item: Item) -> bool:
        """
        Returns:
            if an item will fit on a level above this shelf.
        """
        return self.vertical_offset + self.height + item.height <= self.bin.height

    def _add(self, item: Item) -> None:
        """
        Adds an item to this shelf.
        """
        item._x = self.packed_width
        item._y = self.vertical_offset
        self.bin._add(item)
        self.items.append(item)

        if item.height > self.height:
            self.height = item.height

        self.packed_width += item.width
        self.available_width -= item.width

    def _close(self) -> None:
        """
        Closes this shelf and locks its height.
        """
        self.is_open = False

    def packed_area(self) -> float:
        """
        Returns:
            the amount of area that has been packed into this shelf
        """
        if self.is_open:
            return self.packed_width * self.height
        else:
            return self.width * self.height


# Shelf Packing Algorithms


class ShelfPackingAlgorithm(PackingAlgorithm):
    """
    A class representing a shelf-based packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        super().__init__(bin_width, bin_height)
        self._bins: list[ShelvedBin] = [ShelvedBin(width=bin_width, height=bin_height)]

    def pack(self, item: Item, allow_rotation: bool = True) -> None:
        """
        Packs a rectangular item using a shelf-based algorithm.

        Args:
            item: the item to pack
            allow_rotation: whether to allow rotating the item
        """
        raise NotImplementedError()

    def pack_all(self,
                 items: list[Item],
                 sorting: Optional[SortingFunction] = sort_by_short_side(True),
                 allow_rotation: bool = True) -> None:
        """
        Packs a list of rectangular items into bins.

        Args:
            items: the list of items
            sorting: a sorting function to use, otherwise None.
            allow_rotation: whether to allow rotating the items
        """
        PackingAlgorithm.pack_all(self, items, sorting, allow_rotation=allow_rotation)

    def pack_oriented(self,
                      item: Item,
                      orient_vertically: bool = True) -> None:
        """
        Packs a rectangular item oriented in a particular way using this algorithm.

        Args:
            item: the item to pack
            orient_vertically: whether to orient the item vertically
        """
        if item.is_horizontal() != orient_vertically:
            item._rotate()
        self.pack(item, allow_rotation=False)

    def pack_all_oriented(self,
                          items: list[Item],
                          sorting: Optional[SortingFunction] = sort_by_long_side(descending=True),
                          orient_vertically: bool = True) -> None:
        """
        Packs a list of rectangular items into bins oriented in a particular way.

        Args:
            items: the list of items
            sorting: a sorting function to use, otherwise None.
            orient_vertically: whether to orient the items vertically
        """
        if sorting:
            items = sorting(items)

        for item in items:
            self.pack_oriented(item, orient_vertically=orient_vertically)


class NextFit(ShelfPackingAlgorithm):
    """
    A class representing a shelf-based, next-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, next-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """

        super().__init__(bin_width, bin_height)
        self._shelf = self._bins[-1].open_shelf

    def pack(self, item: Item, allow_rotation: bool = True) -> None:
        """
        Packs a rectangular item using a shelf-based next-fit algorithm.

        Args:
            item: the item to pack
            allow_rotation: whether to allow rotating the item
        """
        if allow_rotation and (item.is_vertical() != (item.long_side <= self._shelf.height)):
            item._rotate()

        if self._shelf._fits(item):
            self._shelf._add(item)
            return

        if self._shelf._fits_above(item):
            self._shelf = self._bins[-1]._add_shelf()
        else:
            self._bins.append(ShelvedBin(width=self._bin_width, height=self._bin_height))
            self._shelf = self._bins[-1].open_shelf

        if allow_rotation and item.is_vertical():
            item._rotate()
        self._shelf._add(item)


class FirstFit(ShelfPackingAlgorithm):
    """
    A class representing a shelf-based, first-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, first-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """

        super().__init__(bin_width, bin_height)
        self._shelf = self._bins[-1].open_shelf

    def pack(self, item: Item, allow_rotation: bool = True) -> None:
        """
        Packs a rectangular item using a shelf-based first-fit algorithm.

        Args:
            item: the item to pack
            allow_rotation: whether to allow rotating the item
        """

        def _add_to_bin() -> bool:
            for shelf in bin:
                if allow_rotation and (item.is_vertical() != (item.long_side <= shelf.height)):
                    item._rotate()

                if shelf._fits(item):
                    shelf._add(item)
                    return True

            if shelf._fits_above(item):  # noqa W0631, each bin has at least 1 shelf
                if allow_rotation and item.is_vertical():
                    item._rotate()
                bin._add_shelf()._add(item)
                return True

            return False

        for bin in self._bins:
            if _add_to_bin():
                break
        else:
            self._bins.append(ShelvedBin(width=self._bin_width, height=self._bin_height))

            if allow_rotation and item.is_vertical():
                item._rotate()
            self._bins[-1].open_shelf._add(item)


class ScoredFit(ShelfPackingAlgorithm):
    """
    A class representing a shelf-based, scored-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float, scoring_function: Callable[[Shelf, Item], float]):
        """
        A class representing a shelf-based, scored-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
            scoring_function: the scoring function to compare shelf candidates
        """
        super().__init__(bin_width, bin_height)
        self._shelf = self._bins[-1].open_shelf
        self._scoring_func = scoring_function

    def pack(self, item: Item, allow_rotation: bool = True) -> None:
        """
        Packs a rectangular item using a shelf-based, scored-fit algorithm.

        Args:
            item: the item to pack
            allow_rotation: whether to allow rotating the item
        """

        def _score(shelf, obj) -> None:
            nonlocal best_shelf, best_score, best_orientation

            score = self._scoring_func(shelf, obj)
            if score > best_score:
                best_shelf = shelf
                best_score = score
                best_orientation = item.is_rotated()

        def _add_to_bin() -> None:
            for shelf in bin:
                if allow_rotation and (item.is_vertical() != (item.long_side <= shelf.height)):
                    item._rotate()

                if shelf._fits(item):
                    _score(shelf, item)

            # pylint: disable-next=undefined-loop-variable
            if best_shelf is None and shelf._fits_above(item):  # noqa W0631, each bin has at least 1 shelf
                shelf = bin._add_shelf()

                if allow_rotation and item.is_vertical():
                    item._rotate()

                _score(shelf, item)

        best_shelf = None
        best_score = float("-inf")
        best_orientation = False  # un-rotated

        for bin in self._bins:
            _add_to_bin()

        if best_shelf is None:
            self._bins.append(ShelvedBin(width=self._bin_width, height=self._bin_height))

            if allow_rotation and item.is_vertical():
                item._rotate()

            best_shelf = self._bins[-1].open_shelf
            best_orientation = item.is_rotated()

        if item.is_rotated() != best_orientation:
            item._rotate()

        best_shelf._add(item)


class BestWidthFit(ScoredFit):
    """
    A class representing a shelf-based, best-width-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, best-width-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        super().__init__(bin_width, bin_height, lambda shelf, obj: obj.width - shelf.available_width)


class WorstWidthFit(ScoredFit):
    """
    A class representing a shelf-based, worst-width-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, best-width-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        super().__init__(bin_width, bin_height, lambda shelf, obj: shelf.available_width - obj.width)


class BestHeightFit(ScoredFit):
    """
    A class representing a shelf-based, best-height-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, best-height-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        super().__init__(bin_width, bin_height, lambda shelf, obj: obj.height - shelf.available_height)


class WorstHeightFit(ScoredFit):
    """
    A class representing a shelf-based, worst-height-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, worst-height-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        super().__init__(bin_width, bin_height, lambda shelf, obj: shelf.available_height - obj.height)


class BestAreaFit(ScoredFit):
    """
    A class representing a shelf-based, best-area-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, best-area-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        super().__init__(bin_width, bin_height,
                         lambda shelf, obj: (shelf.packed_width + obj.width) * max(obj.height, shelf.height))


class WorstAreaFit(ScoredFit):
    """
    A class representing a shelf-based, worst-area-fit packing algorithm.

    Args:
        bin_width: the maximum width of the packed bin
        bin_height: the maximum height of the packed bin
    """

    def __init__(self, bin_width: float, bin_height: float):
        """
        A class representing a shelf-based, worst-area-fit packing algorithm.

        Args:
            bin_width: the maximum width of the packed bin
            bin_height: the maximum height of the packed bin
        """
        super().__init__(bin_width, bin_height,
                         lambda shelf, obj: -(shelf.packed_width + obj.width) * max(obj.height, shelf.height))

# pylint: enable=redefined-builtin

File: src/goopylib/maths/packing/Shelf.cpp
#include "Shelf.h"

#include <cfloat>
#include <utility>

#include "src/config.h"

#if (GP_LOG_SHELF_PACKING != true) and (GP_LOG_SHELF_PACKING <= GP_LOGGING_LEVEL)
#undef GP_LOGGING_LEVEL
#define GP_LOGGING_LEVEL GP_LOG_SHELF_PACKING
#endif

#if !GP_VALUE_CHECK_SHELF_PACKING
#undef GP_VALUE_CHECKING
#undef GP_TYPE_CHECKING
#undef GP_ERROR_CHECKING
#endif

#include "src/goopylib/debug/LogMacros.h"
#include "src/goopylib/debug/Error.h"


// Shelf Class
namespace gp::packing::shelf {
    Shelf::Shelf(float verticalOffset, Bin &bin)
            : m_Bin(bin),
              m_VerticalOffset(verticalOffset),
              m_Width(bin.getWidth()),
              m_AvailableWidth(bin.getWidth()) {

    }

    std::string Shelf::toString() const {
        return strformat("Shelf(offset=%f, packed=%f) with %i items", m_VerticalOffset, m_PackedWidth, m_Items.size());
    }

    float Shelf::packedArea() const {
        if (m_IsOpen) {
            return m_PackedWidth * m_Height;
        }
        else {
            return m_Width * m_Height;
        }
    }

    float Shelf::getWidth() const {
        return m_Width;
    }

    float Shelf::getHeight() const {
        return m_Height;
    }

    float Shelf::getVerticalOffset() const {
        return m_VerticalOffset;
    }

    float Shelf::getPackedWidth() const {
        return m_PackedWidth;
    }

    float Shelf::getAvailableWidth() const {
        return m_AvailableWidth;
    }

    bool Shelf::isOpen() const {
        return m_IsOpen;
    }

    bool Shelf::fits(const Ref<Item> &item) const {
        if (m_IsOpen) {
            return item->getWidth() <= m_AvailableWidth and
                   m_VerticalOffset + item->getHeight() <= m_Bin.getHeight();
        }
        else {
            return item->getWidth() <= m_AvailableWidth and item->getHeight() <= m_Height;
        }
    }

    bool Shelf::fitsAbove(const Ref<Item> &item) const {
        return m_VerticalOffset + m_Height + item->getHeight() <= m_Bin.getHeight();
    }

    void Shelf::add(const Ref<Item> &item) {
        item->m_X = m_PackedWidth;
        item->m_Y = m_VerticalOffset;
        m_Bin.add(item);
        m_Items.push_back(item);

        if (item->getHeight() > m_Height) {
            m_Height = item->getHeight();
        }

        m_PackedWidth += item->getWidth();
        m_AvailableWidth -= item->getWidth();
    }

    void Shelf::close() {
        m_IsOpen = false;
    }

    std::ostream &operator<<(std::ostream &os, const Shelf &shelf) {
        os << shelf.toString();
        return os;
    }
}

// Shelved Bin Class
namespace gp::packing {
    ShelvedBin::ShelvedBin(float width, float height)
            : Bin(width, height),
              m_OpenShelf(Ref<shelf::Shelf>(new shelf::Shelf(0, *this))) {
        m_Shelves.push_back(m_OpenShelf);
    }

    Ref<shelf::Shelf> ShelvedBin::addShelf() {
        m_OpenShelf->close();
        m_OpenShelf = Ref<shelf::Shelf>(new shelf::Shelf(m_OpenShelf->m_VerticalOffset + m_OpenShelf->m_Height, *this));
        m_Shelves.push_back(m_OpenShelf);

        return m_OpenShelf;
    }

    float ShelvedBin::packingRatio() const {
        if (m_ID == Bin::s_Bins - 1) { // this is the latest bin
            float sum = 0;
            float area = 0;
            for (const auto &item: m_Items) {
                sum += item->area();
            }
            for (const auto &shelf: m_Shelves) {
                area += shelf->packedArea();
            }

            return sum / area;
        }
        return Bin::packingRatio();
    }

    Ref<shelf::Shelf> ShelvedBin::getOpenShelf() {
        return m_OpenShelf;
    }

    std::vector<Ref<shelf::Shelf>> ShelvedBin::getShelves() {
        std::vector<Ref<shelf::Shelf>> copy = m_Shelves;
        return copy;
    }

    std::vector<Ref<shelf::Shelf>>::const_iterator ShelvedBin::begin() const {
        return m_Shelves.begin();
    }

    std::vector<Ref<shelf::Shelf>>::const_iterator ShelvedBin::end() const {
        return m_Shelves.end();
    }
}

// Shelf Packing Algorithms
namespace gp {
    namespace packing {
        namespace shelf {
            ShelfPackingAlgorithm::ShelfPackingAlgorithm(float binWidth, float binHeight)
                    : PackingAlgorithm(binWidth, binHeight) {
                m_Bins.push_back(Ref<ShelvedBin>(new ShelvedBin(binWidth, binHeight)));
            }

            void ShelfPackingAlgorithm::pack(const Ref<Item> &item, bool allowRotation) {
                GP_RUNTIME_ERROR("ShelfPackingAlgorithm::pack() unimplemented");
            }

            void ShelfPackingAlgorithm::packAll(std::vector<Ref<Item>> items,
                                                bool allowRotation,
                                                const SortingFunction &sortingFunction) {
                if (sortingFunction) {
                    items = sortingFunction(items);
                }
                for (const auto &item: items) {
                    pack(item, allowRotation);
                }
            }

            void ShelfPackingAlgorithm::packOriented(const Ref<Item> &item, bool orientVertically) {
                if (item->isHorizontal() == orientVertically) {
                    item->rotate();
                }
                pack(item, false);
            }

            void ShelfPackingAlgorithm::packAllOriented(std::vector<Ref<Item>> items,
                                                        bool orientVertically,
                                                        const SortingFunction &sortingFunction) {
                if (sortingFunction) {
                    items = sortingFunction(items);
                }
                for (auto &item: items) {
                    packOriented(item, orientVertically);
                }
            }

            std::vector<Ref<ShelvedBin>> ShelfPackingAlgorithm::bins() const {
                std::vector<Ref<ShelvedBin>> copy = m_Bins;
                return copy;
            }

            NextFit::NextFit(float binWidth, float binHeight)
                    : ShelfPackingAlgorithm(binWidth, binHeight),
                      m_Shelf(m_Bins.back()->m_OpenShelf) {
            }

            void NextFit::pack(const Ref<Item> &item, bool allowRotation) {
                if (allowRotation and (item->isVertical() != (item->getLongSide() <= m_Shelf->getHeight()))) {
                    item->rotate();
                }

                if (m_Shelf->fits(item)) {
                    m_Shelf->add(item);
                    return;
                }

                if (m_Shelf->fitsAbove(item)) {
                    m_Shelf = m_Bins.back()->addShelf();
                }
                else {
                    m_Bins.push_back(Ref<ShelvedBin>(new ShelvedBin(m_BinWidth, m_BinHeight)));
                    m_Shelf = m_Bins.back()->m_OpenShelf;
                }

                if (allowRotation and item->isVertical()) {
                    item->rotate();
                }
                m_Shelf->add(item);
            }

            FirstFit::FirstFit(float binWidth, float binHeight)
                    : ShelfPackingAlgorithm(binWidth, binHeight) {
            }

            void FirstFit::pack(const Ref<Item> &item, bool allowRotation) {
                for (const auto &bin: m_Bins) {
                    for (const auto &shelf: *bin) {
                        if (allowRotation and (item->isVertical() != (item->getLongSide() <= shelf->getHeight()))) {
                            item->rotate();
                        }

                        if (shelf->fits(item)) {
                            shelf->add(item);
                            return;
                        }
                    }

                    if (bin->m_OpenShelf->fitsAbove(item)) {
                        if (allowRotation and item->isVertical()) {
                            item->rotate();
                        }
                        bin->addShelf()->add(item);
                        return;
                    }
                }
                // code only reaches here if item has not been added to a shelf
                m_Bins.push_back(Ref<ShelvedBin>(new ShelvedBin(m_BinWidth, m_BinHeight)));

                if (allowRotation and item->isVertical()) {
                    item->rotate();
                }
                const auto &newBin = m_Bins.back();
                const auto &newShelf = newBin->m_OpenShelf;
                newShelf->add(item);
            }

            ScoredFit::ScoredFit(float binWidth, float binHeight, ScoringFunction scoringFunction)
                    : ShelfPackingAlgorithm(binWidth, binHeight),
                      m_ScoringFunction(std::move(scoringFunction)) {
            }

            void ScoredFit::pack(const Ref<Item> &item, bool allowRotation) {
                Ref<Shelf> bestShelf = nullptr;
                float bestScore = -FLT_MAX;
                bool bestOrientation = false;  // un-rotated

                for (const auto &bin: m_Bins) {
                    for (const auto &shelf: *bin) {
                        if (allowRotation and (item->isVertical() != (item->getLongSide() <= shelf->getHeight()))) {
                            item->rotate();
                        }

                        if (shelf->fits(item)) {
                            float score = m_ScoringFunction(shelf, item);
                            if (score > bestScore) {
                                bestShelf = shelf;
                                bestScore = score;
                                bestOrientation = item->isRotated();
                            }
                        }
                    }

                    if (bestShelf == nullptr and bin->m_OpenShelf->fitsAbove(item)) {
                        auto shelf = bin->addShelf();

                        float score = m_ScoringFunction(shelf, item);
                        if (score > bestScore) {
                            bestShelf = shelf;
                            bestScore = score;
                            bestOrientation = item->isRotated();
                        }
                    }
                }

                if (bestShelf == nullptr) {
                    m_Bins.push_back(Ref<ShelvedBin>(new ShelvedBin(m_BinWidth, m_BinHeight)));

                    if (allowRotation and item->isVertical()) {
                        item->rotate();
                    }

                    bestShelf = m_Bins.back()->m_OpenShelf;
                    bestOrientation = item->isRotated();
                }

                if (item->isRotated() != bestOrientation) {
                    item->rotate();
                }
                bestShelf->add(item);
            }

            BestWidthFit::BestWidthFit(float binWidth, float binHeight)
                    : ScoredFit(binWidth, binHeight,
                                [](const Ref<Shelf> &shelf, const Ref<Item> &obj) {
                                    return obj->getWidth() - shelf->getAvailableWidth();
                                }) {
            }

            WorstWidthFit::WorstWidthFit(float binWidth, float binHeight)
                    : ScoredFit(binWidth, binHeight,
                                [](const Ref<Shelf> &shelf, const Ref<Item> &obj) {
                                    return shelf->getAvailableWidth() - obj->getWidth();
                                }) {
            }

            BestHeightFit::BestHeightFit(float binWidth, float binHeight)
                    : ScoredFit(binWidth, binHeight,
                                [](const Ref<Shelf> &shelf, const Ref<Item> &obj) {
                                    return obj->getHeight() - shelf->getHeight();
                                }) {
            }

            WorstHeightFit::WorstHeightFit(float binWidth, float binHeight)
                    : ScoredFit(binWidth, binHeight,
                                [](const Ref<Shelf> &shelf, const Ref<Item> &obj) {
                                    return shelf->getHeight() - obj->getHeight();
                                }) {
            }

            BestAreaFit::BestAreaFit(float binWidth, float binHeight)
                    : ScoredFit(binWidth, binHeight,
                                [](const Ref<Shelf> &shelf, const Ref<Item> &obj) {
                                    return (shelf->getPackedWidth() + obj->getWidth()) *
                                           max(obj->getHeight(), shelf->getHeight());
                                }) {
            }

            WorstAreaFit::WorstAreaFit(float binWidth, float binHeight)
                    : ScoredFit(binWidth, binHeight,
                                [](const Ref<Shelf> &shelf, const Ref<Item> &obj) {
                                    return -(shelf->getPackedWidth() + obj->getWidth()) *
                                           max(obj->getHeight(), shelf->getHeight());
                                }) {
            }
        }
    }
}